rayon-core/src/future.rs

```diff
@@ -0,0 +1,139 @@
+#![allow(missing_docs)]
+
+use crate::ThreadPool;
+use crate::{spawn, spawn_fifo};
+use crate::{Scope, ScopeFifo};
+
+use std::future::Future;
+use std::mem;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll, Waker};
+
+use job::JobResult;
+use unwind;
+
+struct RayonFuture<T> {
+    state: Arc<Mutex<State<T>>>,
+}
+
+struct RayonFutureJob<T> {
+    state: Arc<Mutex<State<T>>>,
+}
+
+struct State<T> {
+    result: JobResult<T>,
+    waker: Option<Waker>,
+}
+
+fn new<T>() -> (RayonFuture<T>, RayonFutureJob<T>) {
+    let state = Arc::new(Mutex::new(State {
+        result: JobResult::None,
+        waker: None,
+    }));
+    (
+        RayonFuture {
+            state: state.clone(),
+        },
+        RayonFutureJob { state },
+    )
+}
+
+impl<T> Future for RayonFuture<T> {
+    type Output = T;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let mut guard = self.state.lock().expect("rayon future lock");
+        match mem::replace(&mut guard.result, JobResult::None) {
+            JobResult::None => {
+                guard.waker = Some(cx.waker().clone());
+                Poll::Pending
+            }
+            JobResult::Ok(x) => Poll::Ready(x),
+            JobResult::Panic(p) => {
+                drop(guard); // don't poison the lock
+                unwind::resume_unwinding(p);
```

Comment:
This is usually how we propagate panics, but maybe the future should yield `Result<T, ...>` instead?

Reply:
I think it would be best to let Rayon's `panic_handler` handle the actual panic, but also panic here with something like `panic!("the spawned task has panicked")` rather than resuming with the original one.

If one were to retrieve the result of a spawned task without using futures, they would probably create a channel and send the result through it. Then, if the task panics, the sender side of the channel gets dropped, thus disconnecting it. If one attempts to receive the result from the receiver side of the channel, the `receiver.recv()` call panics because the channel is disconnected.

So that way `spawn_future` would closely match the behavior of `spawn` + using a channel to retrieve the result.

Over the past few months I did a lot of exploration with panic propagation strategies in asynchronous contexts and talked to people about their use cases. In the end, I figured the generally best way of handling panics is to pass them to the panic handler and raise a new panic whenever the result of a failed task is polled.

Reply:
I'd be OK with that too. I was thinking of `Result` as an analogy from `std::thread::spawn` + `JoinHandle::join` to `rayon::spawn_future` + `Future::poll`.